MarshallingAttributeAnalyzers throws InvalidCastException

### Description

NativeMarshallingAttributeAnalyzer and CustomMarshallerAttributeAnalyzer are throwing exceptions when building the runtime repo.

### Reproduction Steps

I upgraded an existing ADO pipeline to use .NET 7 RC 1 and it started to fail.
I have no idea what configuration in repo is relevant for repro, I'll be happy to share more details and link to pipeline as needed.

### Expected behavior

No errors

### Actual behavior

##[error]CSC(0,0): Error AD0001: Analyzer 'Microsoft.Interop.Analyzers.NativeMarshallingAttributeAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.NonErrorNamedTypeSymbol' to type 'Microsoft.CodeAnalysis.IMethodSymbol'.'.
     7>CSC : error AD0001: Analyzer 'Microsoft.Interop.Analyzers.NativeMarshallingAttributeAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.NonErrorNamedTypeSymbol' to type 'Microsoft.CodeAnalysis.IMethodSymbol'.'. [D:\a\_work\10\s\src\Product\UITester.Mac\Microsoft.Mashup.Client.UITester.Mac.csproj]
##[error]CSC(0,0): Error AD0001: Analyzer 'Microsoft.Interop.Analyzers.CustomMarshallerAttributeAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.NonErrorNamedTypeSymbol' to type 'Microsoft.CodeAnalysis.IMethodSymbol'.'.
     7>CSC : error AD0001: Analyzer 'Microsoft.Interop.Analyzers.CustomMarshallerAttributeAnalyzer' threw an exception of type 'System.InvalidCastException' with message 'Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel.NonErrorNamedTypeSymbol' to type 'Microsoft.CodeAnalysis.IMethodSymbol'.'. [D:\a\_work\10\s\src\Product\UITester.Mac\Microsoft.Mashup.Client.UITester.Mac.csproj]

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

Seems similar to https://github.com/dotnet/runtime/issues/72622 , but this time with InvalidCastException instead of InvalidOperationException.

--- a/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/SyntaxExtensions.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/SyntaxExtensions.cs
@@ -33,7 +33,17 @@ namespace Microsoft.Interop.Analyzers
                 switch (attributeTarget.Identifier.Kind())
                 {
                     case SyntaxKind.ReturnKeyword:
-                        return ((IMethodSymbol)targetSymbol).GetReturnTypeAttributes().First(attributeSyntaxLocationMatches);
+                        if (targetSymbol is IMethodSymbol method)
+                        {
+                            // Sometimes an attribute is put on a symbol that is nested within the containing symbol.
+                            // For example, the ContainingSymbol for an AttributeSyntax on a local function have a ContainingSymbol of the method.
+                            // Since this method is internal and the callers don't care about attributes on local functions,
+                            // we just allow this method to return null in those cases.
+                            return method.GetReturnTypeAttributes().FirstOrDefault(attributeSyntaxLocationMatches);
+                        }
+                        // An attribute on the return value of a delegate type's Invoke method has a ContainingSymbol of the delegate type.
+                        // We don't care about the attributes in this case for the callers, so we'll just return null.
+                        return null;
                     case SyntaxKind.AssemblyKeyword:
                         return targetSymbol.ContainingAssembly.GetAttributes().First(attributeSyntaxLocationMatches);
                     case SyntaxKind.ModuleKeyword:
@@ -43,7 +53,8 @@ namespace Microsoft.Interop.Analyzers
                 }
             }
             // Sometimes an attribute is put on a symbol that is nested within the containing symbol.
-            // For example, the ContainingSymbol for an AttributeSyntax on a parameter have a ContainingSymbol of the method.
+            // For example, the ContainingSymbol for an AttributeSyntax on a parameter have a ContainingSymbol of the method
+            // and an AttributeSyntax on a local function have a ContainingSymbol of the containing method.
             // Since this method is internal and the callers don't care about attributes on parameters, we just allow
             // this method to return null in those cases.
             return targetSymbol.GetAttributes().FirstOrDefault(attributeSyntaxLocationMatches);

--- a/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/SyntaxExtensions.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/SyntaxExtensions.cs
@@ -36,7 +36,7 @@ namespace Microsoft.Interop.Analyzers
                         if (targetSymbol is IMethodSymbol method)
                         {
                             // Sometimes an attribute is put on a symbol that is nested within the containing symbol.
-                            // For example, the ContainingSymbol for an AttributeSyntax on a local function have a ContainingSymbol of the method.
+                            // For example, the ContainingSymbol for an AttributeSyntax on a local function has a ContainingSymbol of the method.
                             // Since this method is internal and the callers don't care about attributes on local functions,
                             // we just allow this method to return null in those cases.
                             return method.GetReturnTypeAttributes().FirstOrDefault(attributeSyntaxLocationMatches);
@@ -53,8 +53,8 @@ namespace Microsoft.Interop.Analyzers
                 }
             }
             // Sometimes an attribute is put on a symbol that is nested within the containing symbol.
-            // For example, the ContainingSymbol for an AttributeSyntax on a parameter have a ContainingSymbol of the method
-            // and an AttributeSyntax on a local function have a ContainingSymbol of the containing method.
+            // For example, the ContainingSymbol for an AttributeSyntax on a parameter has a ContainingSymbol of the method
+            // and an AttributeSyntax on a local function has a ContainingSymbol of the containing method.
             // Since this method is internal and the callers don't care about attributes on parameters, we just allow
             // this method to return null in those cases.
             return targetSymbol.GetAttributes().FirstOrDefault(attributeSyntaxLocationMatches);

--- a/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/NativeMarshallingAttributeAnalyzerTests.cs
+++ b/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/NativeMarshallingAttributeAnalyzerTests.cs
@@ -272,8 +272,18 @@ namespace LibraryImportGenerator.UnitTests
                 {
                     void Foo([MarshalAs(UnmanagedType.I4)] int i)
                     {
+                        [return:MarshalAs(UnmanagedType.I4)]
+                        [SkipLocalsInit]
+                        static int Local()
+                        {
+                            return 0;
+                        }
                     }
                 }
+
+                [return:MarshalAs(UnmanagedType.I4)]
+                delegate int Y();
+
                 """;
 
             await VerifyCS.VerifyAnalyzerAsync(source);

